block/internal/executing/executor.go

```diff
@@ -0,0 +1,643 @@
+	}
+
+	if delay > 0 {
+		e.logger.Info().Dur("delay", delay).Msg("waiting to start block production")
+		select {
+		case <-e.ctx.Done():
+			return
+		case <-time.After(delay):
+		}
+	}
+
+	ticker := time.NewTicker(e.config.Node.BlockTime.Duration)
+	defer ticker.Stop()
+
+	txsAvailable := false
+
+	for {
+		select {
+		case <-e.ctx.Done():
+			return
+		case <-ticker.C:
+			if e.config.Node.LazyMode && !txsAvailable {
+				// In lazy mode, only produce blocks when transactions are available
+				continue
+			}
+
+			if err := e.produceBlock(); err != nil {
+				e.logger.Error().Err(err).Msg("failed to produce block")
+				// Continue execution loop even on error
+			}
+
+			txsAvailable = false
+		case <-e.txNotifyCh:
+			txsAvailable = true
+		}
+	}
+}
+
+// produceBlock creates, validates, and stores a new block
+func (e *Executor) produceBlock() error {
+	start := time.Now()
+	defer func() {
+		if e.metrics.OperationDuration["block_production"] != nil {
+			duration := time.Since(start).Seconds()
+			e.metrics.OperationDuration["block_production"].Observe(duration)
+		}
+	}()
+
+	ctx := context.Background()
+	currentState := e.GetLastState()
+	newHeight := currentState.LastBlockHeight + 1
+
+	e.logger.Debug().Uint64("height", newHeight).Msg("producing block")
+
+	// Check pending limits
+	if e.config.Node.MaxPendingHeadersAndData > 0 {
+		pendingHeaders := e.cache.NumPendingHeaders()
+		pendingData := e.cache.NumPendingData()
+		if pendingHeaders >= e.config.Node.MaxPendingHeadersAndData ||
+			pendingData >= e.config.Node.MaxPendingHeadersAndData {
+			e.logger.Warn().
+				Uint64("pending_headers", pendingHeaders).
+				Uint64("pending_data", pendingData).
+				Uint64("limit", e.config.Node.MaxPendingHeadersAndData).
+				Msg("pending limit reached, skipping block production")
+			return nil
+		}
+	}
+
+	// Get batch from sequencer
+	batchData, err := e.retrieveBatch(ctx)
+	if errors.Is(err, common.ErrNoBatch) {
+		e.logger.Debug().Msg("no batch available")
+		return nil
+	} else if errors.Is(err, common.ErrNoTransactionsInBatch) {
+		e.logger.Debug().Msg("no transactions in batch")
+	} else if err != nil {
+		return fmt.Errorf("failed to retrieve batch: %w", err)
+	}
+
+	// Create block
+	header, data, err := e.createBlock(ctx, newHeight, batchData)
+	if err != nil {
+		return fmt.Errorf("failed to create block: %w", err)
+	}
+
+	// Apply block to get new state
+	newState, err := e.applyBlock(ctx, header.Header, data)
+	if err != nil {
+		return fmt.Errorf("failed to apply block: %w", err)
+	}
+
+	// Sign the header
+	signature, err := e.signHeader(header.Header)
+	if err != nil {
+		return fmt.Errorf("failed to sign header: %w", err)
+	}
+	header.Signature = signature
+
+	// Validate block
+	if err := e.validateBlock(currentState, header, data); err != nil {
+		return fmt.Errorf("failed to validate block: %w", err)
+	}
+
+	// Save block
+	if err := e.store.SaveBlockData(ctx, header, data, &signature); err != nil {
```

Comment:
i think we have issues in this flow, since we applied but may have not saved we have been ending up in a corrupted state